src/NHibernate/Criterion/SqlFunctionProjection.cs

```diff
@@ -107,10 +114,17 @@ private static SqlString GetProjectionArgument(ICriteriaQuery criteriaQuery, ICr
 		}
 
 		public override IType[] GetTypes(ICriteria criteria, ICriteriaQuery criteriaQuery)
+		{
+			return new IType[] {GetReturnType(criteria, criteriaQuery)};
+		}
+
+		private IType GetReturnType(ICriteria criteria, ICriteriaQuery criteriaQuery)
 		{
 			ISQLFunction sqlFunction = GetFunction(criteriaQuery);
-			IType type = sqlFunction.ReturnType(returnType, criteriaQuery.Factory);
-			return new IType[] {type};
+
+			var resultType = returnType ?? returnTypeProjection?.GetTypes(criteria, criteriaQuery).FirstOrDefault();
```

Comment:
If there are more than one type returned, should it be silently ignored? (Shouldn't `SingleOrDefault()`be called here?)

And all the null coalesce, OrDefault stuff combined with yielding an `IType[]` of one element which may as a result be `null` does not look good. Digging on the `GetTypes` usage, it seems it ends up being used by calling methods on the elements it contains without checking if they are `null` or not. (See `CriteriaLoader.GetResultRow`.)
So either it should yield an empty type array when no type are determined if such an empty array can be a valid return value, or throw for the sake of failing earlier.
This defect was somehow already there, since nothing was checking the sqlFunction return type was not `null`. But it looks worsen by the change.

Reply:
> If there are more than one type returned, should it be silently ignored? (Shouldn't SingleOrDefault() be called here?)

Well I don't know. I made it the way I saw somewhere else in Criteria handling. IMHO it's not important until multi-column support is added to `ISQLFunction.ReturnType`

> And all the null coalesce, OrDefault stuff combined with yielding an IType[] of one element which may as a result be null does not look good. ... it seems it ends up being used by calling methods

Not sure I properly understand it. But this logic ends up in `sqlFunction.ReturnType` which expects single `IType` and returns single `IType`.  And I see no restrictions why `null` should be forbidden for `sqlFunction.ReturnType` - function return type is known beforehand in most cases . So both `returnType` and `returnTypeProjection` should be nullable.

Reply:
The trouble is, `GetTypes` returned array is used as if its content could never be null, without null checks, see `CriteriaLoader.GetResultRow`. So potentially yielding  `new IType [] { null }` looks plain wrong.
The PR introduces more possibilities for this to happen.
I think is should either yield an empty array, or throw. It should not yield an array containing null. Or is there something I have missed which this returned value to be actually supported?

Reply:
No but empty type collection also doesn't make much sense for using in Select projection... Ok - I've added null handling just in case.